### ISSUE TITLE
loss function for multi-label problem

### ISSUE THREAD

**Issue by kingfengji** [fchollet/keras#2166]
Hi there, How to choose loss function for multi-label problem

it's different from multi-class output, the former output is a 0/1 vector with multiple ones, whereas the multi-class output is a single one-hot vector.

Thanks

**Comment by kingfengji** [comment]
the multi-class entropy thing is for multi-class problem, I suppose?

**Comment by NasenSpray** [comment]
> the multi-class entropy thing is for multi-class problem, I suppose?

Yes. What you want is `binary_crossentropy`

**Comment by kingfengji** [comment]
@NasenSpray binary_crossentropy is for multiclass, but not multilabel, right?

**Comment by NasenSpray** [comment]
`categorical_crossentropy`: 1-of-N (one-hot)  
`binary_crossentropy`: 1-or-more 0/1 labels

**Comment by keunwoochoi** [comment]
MSE/MAE also work, binary crossentropy would be preferred in general though.

**Comment by alyato** [comment]
@kingfengji  i'm doing multi-label and do you share your code how to do it. Or Do you give one simple example how to implement multi-label classification.Thanks

**Comment by suraj-deshmukh** [comment]
@alyato this might help you [multi label image classification](https://github.com/suraj-deshmukh/Multi-Label-Image-Classification)

**Comment by michelleowen** [comment]
If using binary_crossentropy as loss function, does it mean we are minimizing the average of all cross-entropies over all classes?

**Comment by keunwoochoi** [comment]
I believe so.

**Comment by lipeipei31** [comment]
@keunwoochoi Could you explain why binary crossentropy is preferred for multi-label classification? I thought  binary crossentropy was only for binary classification where y label is only 0 or 1. Now that the y label is in the format of [1,0,1,0,1..], do you know how the loss is calculated with binary crossentropy?

**Comment by 1064950364** [comment]
Thanks,My last layer is softmax layer. I use 'binary_crossentropy' as loss and I get 99% accuracy, while I use other  loss function, I get only 10% accuracy.
I want to know how the accuracy is calculated?

**Comment by 1064950364** [comment]
I get high accuracy,but when I see the predict labels. I find the labels are all-zeros.

**Comment by keunwoochoi** [comment]
@1064950364 Yes, that's the definition of accuracy and that's why accuracy doesn't matter with many multi-label problems. In your true labels, there are so many zeros, right?
@lipeipei31 More precisely, crossentropy is preferred over MAE/MSE. It's bounded, it's loss computation (which is in proportion to the gradient applied) is more plausible. In that case it computes crossetnropy over each output and then compute their average.

**Comment by alyato** [comment]
@1064950364 Do you compare the two different layer in the last layer,softmax or sigmoid.

**Comment by lipeipei31** [comment]
@sarthakahuja11, sounds like you have a multi-output problem where each output is binary or multi-class classification. I think you should have different loss functions for different outputs.

**Comment by sarthakahuja11** [comment]
@lipeipei31  You have identified the problem correctly. So I should choose binary cross entropy for binary-class classification and categorical-cross entropy for multi-class classification? And combine them together afterwards in the same model?

**Comment by lipeipei31** [comment]
@sarthakahuja11 Yes, that's right. And you can easily do that with the keras functional api:
https://keras.io/getting-started/functional-api-guide/#multi-input-and-multi-output-models. And the loss functions can be a list or dictionary if you named the outputs.

**Comment by sarthakahuja11** [comment]
Thanks! @lipeipei31

**Comment by Sonius94** [comment]
hi, 

if binary cross entropy is working in Keras for multi-label problems, will categorical_crossentropy work for multi one-hot encoded classes as well?

My example output is:
[
    [0,0,1,0]
    [0,0,0,1]
    [1,0,0,0]
]

So I have three one hot encoded vectors. For a single on the loss function to choose would be categorical cross entropy. What will Keras do in a case like these?
